Implement Emscripten support

TestingPlant opened the pull request:
This adds support for Emscripten.

This is required to support https://github.com/SanderMertens/flecs/issues/1538

TestingPlant committed:
Disable backtrace for Emscripten

TestingPlant committed:
Add Emscripten build

TestingPlant committed:
Add macros for Emscripten/wasm

SanderMertens commented:
Note that you can build for emscripten with bake if you do
```
bake --target em
```

TestingPlant commented:
> Note that you can build for emscripten with bake if you do
> 
> ```
> bake --target em
> ```

`bake setup --local --target em` doesn't currently compile, and it attempts to generate .so files which aren't supported by Emscripten. I'm currently working on making the drivers build to a static library when on Emscripten and getting everything compiling.

TestingPlant committed:
Update premake files

Requires premake 5.0.0-beta4 or newer which supports Emscripten

TestingPlant committed:
Update Emscripten library prefix/ext

TestingPlant committed:
Compile drivers for Emscripten

TestingPlant committed:
Replace strdup with ut_strdup

TestingPlant committed:
Use -sSIDE_MODULE on emcc

TestingPlant committed:
Fix premake filters

TestingPlant committed:
Use file naming conventions of current target

This removes the global variables used to store file naming conventions,
since we need to be able to use the host file naming conventions to find
drivers and the target file naming conventions for everything else.

SanderMertens commented:
How're things going? Anything I can help with?

TestingPlant committed:
Statically link on Emscripten

TestingPlant committed:
Remove Emscripten driver builds

TestingPlant committed:
Compile bake_util with pic on Emscripten

TestingPlant committed:
Run compiled files on Emscripten

TestingPlant committed:
Re-add MAIN_MODULE

TestingPlant committed:
Enable pthread on Emscripten

TestingPlant committed:
Remove Emscripten from run_premake.sh

TestingPlant committed:
Remove modularize for emcc

TestingPlant committed:
Add ut_proc stubs for Emscripten

TestingPlant committed:
Add single threaded mode to test runner

TestingPlant committed:
Allow test runner to run without processes

TestingPlant committed:
Add more flags to emcc

TestingPlant committed:
Add shebang manually

TestingPlant committed:
Fix return code handling

TestingPlant committed:
Suppress MAIN_MODULE+pthreads warning
